staging/src/k8s.io/apiserver/pkg/cel/library/semver.go

```diff
@@ -0,0 +1,230 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package library
+
+import (
+	"github.com/blang/semver/v4"
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+
+	apiservercel "k8s.io/apiserver/pkg/cel"
+)
+
+// SemVer provides a CEL function library extension for [semver.Version].
+//
+// semver
+//
+// Converts a string to a SemVer or results in an error if the string is not a valid SemVer. Refer
+// to semver.SemVer documentation for information on accepted patterns.
+//
+//	semver(<string>) <SemVer>
+//
+// Examples:
+//
+//	semver('1.0.0') // returns a SemVer
+//	semver('0.1.0-alpha.1') // returns a SemVer
+//	semver('200K') // error
+//	semver('Three') // error
+//	semver('Mi') // error
+//
+// isSemVer
+//
+// Returns true if a string is a valid SemVer. isSemVer returns true if and
+// only if semver does not result in error.
+//
+//	isSemVer( <string>) <bool>
+//
+// Examples:
+//
+//	isSemVer('1.0.0') // returns true
+//	isSemVer('v1.0') // returns true (tolerant parsing)
+//	isSemVer('hello') // returns false
+//
+// Conversion to Scalars:
+//
+//   - major/minor/patch: return the major version number as int64.
+//
+//     <SemVer>.major() <int>
+//
+// Examples:
+//
+// semver("1.2.3").major() // returns 1
+//
+// Comparisons
+//
+//   - isGreaterThan: Returns true if and only if the receiver is greater than the operand
+//
+//   - isLessThan: Returns true if and only if the receiver is less than the operand
+//
+//   - compareTo: Compares receiver to operand and returns 0 if they are equal, 1 if the receiver is greater, or -1 if the receiver is less than the operand
+//
+//
+//     <SemVer>.isLessThan(<semver>) <bool>
+//     <SemVer>.isGreaterThan(<semver>) <bool>
+//     <SemVer>.compareTo(<semver>) <int>
+//
+// Examples:
+//
+// semver("1.2.3").compareTo(semver("1.2.3")) // returns 0
+// semver("1.2.3").compareTo(semver("2.0.0")) // returns -1
+// semver("1.2.3").compareTo(semver("0.1.2")) // returns 1
+
+func SemVer() cel.EnvOption {
+	return cel.Lib(semverLib)
+}
+
+var semverLib = &semverLibType{}
+
+type semverLibType struct{}
+
+func (*semverLibType) LibraryName() string {
+	return "k8s.semver"
+}
+
+var semverLibraryDecls = map[string][]cel.FunctionOpt{
+	"semver": {
+		cel.Overload("string_to_semver", []*cel.Type{cel.StringType}, apiservercel.SemVerType, cel.UnaryBinding((stringToSemVer))),
+	},
+	"isSemVer": {
```

Comment:
nit: is it weird to have `semver` but `isSemVer`?

seems like we should have a matched pair:
* semver / isSemver
* semVer / isSemVer